quine-core/src/main/scala/com/thatdot/quine/graph/NodeActor.scala

```diff
@@ -139,6 +134,73 @@ private[graph] class NodeActor(
     case MergedHere(other) => !mergedIntoHere.contains(other)
   }
 
+  /**
+    * Drop duplicate events, and wrap the Events being kept in an `EventTime`.
+    *
+    * An event is considered a duplicate (and is dropped) if:
+    *  - the same event appears later in the sequence of events, or
+    *  - the event would have no effect when applied to this node.
+    *
+    * This implementation prioritizes avoiding heap allocations.
+    * Since it does a linear search of the later events,
+    * this could be less efficient for very large sequences of events.
+    *
+    * @param events An ordered sequence of events to be de-duplicated.
+    * @param atTimeOverride If defined, the `EventTime` that should be combined with the single event;
+    *                       If empty, the an `EventTime` is obtained using `nextEventTime()`.
+    * @return a sequence of de-duplicated events wrapped in a `WithTime`.
+    *         This result is mutable, so it should not be allowed to escape this context.
+    */
+  def deduplicateEvents(events: Seq[NodeChangeEvent],
```

Comment:
Nit: As a private helper function, this shouldn't be called outside this behavior. I'd recommend:
```suggestion
  private[this] def deduplicateEvents(events: Seq[NodeChangeEvent],
```